presto-main/src/main/java/com/facebook/presto/operator/scalar/BitwiseFunctions.java

```diff
@@ -75,4 +83,234 @@ public static long bitwiseXor(@SqlType(StandardTypes.BIGINT) long left, @SqlType
+    @ScalarFunction("bitwise_left_shift")
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseLeftShiftBigint(@SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        return value << shift;
+    }
+
+    private static long preserveSign(long shiftedValue, long mask, long signedBit)
+    {
+        if ((shiftedValue & signedBit) != 0) {
+            // Preserve the sign in 2's complement format
+            return shiftedValue | ~mask;
+        }
+
+        return shiftedValue & mask;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseRightShiftTinyint(@SqlType(StandardTypes.TINYINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        if (shift == 0) {
+            return value;
+        }
+        return (value & TINYINT_MASK) >>> shift;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.SMALLINT)
+    public static long bitwiseRightShiftSmallint(@SqlType(StandardTypes.SMALLINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        if (shift == 0) {
+            return value;
+        }
+        return (value & SMALLINT_MASK) >>> shift;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.INTEGER)
+    public static long bitwiseRightShiftInteger(@SqlType(StandardTypes.INTEGER) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        if (shift == 0) {
+            return value;
+        }
+        return (value & INTEGER_MASK) >>> shift;
+    }
+
+    @Description("bitwise logical right shift")
+    @ScalarFunction("bitwise_right_shift")
+    @SqlType(StandardTypes.BIGINT)
+    public static long bitwiseRightShiftBigint(@SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            return 0L;
+        }
+        return value >>> shift;
+    }
+
+    @Description("bitwise arithmetic right shift")
+    @ScalarFunction("bitwise_right_shift_arithmetic")
+    @SqlType(StandardTypes.TINYINT)
+    public static long bitwiseRightShiftArithmeticTinyint(@SqlType(StandardTypes.TINYINT) long value, @SqlType(StandardTypes.INTEGER) long shift)
+    {
+        if (shift >= MAX_BITS) {
+            if (value >= 0) {
+                return 0L;
+            }
+            else {
+                return -1L;
+            }
+        }
+        return preserveSign(value, TINYINT_MASK, TINYINT_SIGNED_BIT) >> shift;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 258:266, 274:282, 290:298.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.